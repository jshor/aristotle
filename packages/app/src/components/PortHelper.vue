<template>
  <div
    class="port-helper snappable"
    :class="`port-helper--${orientation}`"
  />
</template>

<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  name: 'PortHelper',
  props: {
  /**
   * Sets the visual orientation for the port snap helper.
   */
    orientation: {
      type: String,
      default: 'horizontal'
    }
  }
})
</script>

<style lang="scss">
.port-helper {
  content: ' ';
  display: block;
  position: absolute;
  background-color: red;

  &--vertical {
    width: 1px;
    left: -1px;
    top: -200vh;
    height: 400vh;
  }

  &--horizontal {
    height: 1px;
    left: -200vh;
    width: 400vh;
  }
}
</style>
