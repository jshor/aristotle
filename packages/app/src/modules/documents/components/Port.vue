<template>
  <div class="port" :title="rotation" :style="{
    transform: `rotate(-${rotation * 90}deg)`
  }">
    <!-- <div class="port__helper port__helper--vertical snappable" />
    <div class="port__helper port__helper--horizontal snappable" /> -->
    <div class="port__handle"  />
      <draggable class="port" @drag="onDrag" @dragStart="dragStart" @dragEnd="dragEnd" :position="draggablePosition" :data-id="id" :class="{ snappable }" v-if="draggable">
        <div class="port__handle" @mousedown="mousedown" @mouseup="mouseup" :class="{ 'port__handle--active': snappable }"  />
      </draggable>
      <slot />
    <!-- </div> -->
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { Component, Prop } from 'vue-property-decorator'
import Draggable from './Draggable.vue'
import { Getter } from '../store/decorators'

@Component({
  components: {
    Draggable
  }
})
export default class Port extends Vue {

  @Getter('documents', 'zoom')
  public zoom: number

  @Prop()
  public id: string

  @Prop({ default: 0 })
  public rotation: any

  @Prop({ default: () => ({
    x: 0, y: 0
  }) })
  public position: any

  @Prop({ default: true })
  public draggable: boolean

  @Prop({ default: 0 })
  public type: number

  @Prop({ default: 0 })
  public orientation: number

  public cloneId: string = 'DRAGGED_PORT'

  public draggablePosition: any = {
    x: 0,
    y: 0
  }

  public absolutePosition: any = {
    x: 0,
    y: 0
  }

  get snappable () {
    const { activePort } = this.$store.state.documents

    return activePort !== null // && activePort.type !== this.type
  }

  setAbsolutePosition () {
    const parent = (this.$parent.$parent.$parent.$parent.$refs.canvas as any).getBoundingClientRect()
    const { x, y } = this.$el.getBoundingClientRect()

    this.absolutePosition = {
      x: x - (parent.x * this.zoom),
      y: y - (parent.y * this.zoom)
    }
  }

  mousedown () {
    this.setAbsolutePosition()
    this.setActivePort({
      id: this.cloneId,
      position: {
        x: 0,
        y: 0
      },
      type: this.type,
      orientation: this.orientation
    })
  }

  mouseup () {
    this.$store.dispatch('setActivePort', null)
  }

  dragStart ({ position }) {
    this.draggablePosition = {}
    this.onDrag(position)

    this.$store.dispatch('connect', {
      source: this.id,
      target: this.cloneId
    })
  }

  onDrag ({ position }) {
    if (!position) return
    this.$store.dispatch('updatePortPositions', {
      [this.cloneId]: {
        position: {
          x: position.x + this.absolutePosition.x,
          y: position.y + this.absolutePosition.y
        },
        id: this.cloneId,
        type: this.type,
        orientation: this.orientation,
        rotation: 0
      }
    })
  }

  dragEnd ({ position, snappedId }) {
    this.draggablePosition = {
      x: 0,
      y: 0
    }

    this.$store.dispatch('disconnect', {
      source: this.id,
      target: this.cloneId
    })
    this.$store.dispatch('connect', {
      source: this.id,
      target: snappedId
    })
    this.$store.dispatch('setActivePort', null)
  }

  setActivePort (port) {
    this.$store.dispatch('setActivePort', port)
  }
}
</script>

<style lang="scss">
.dr {
  background:blue;
}
.port {
  width: 0;
  height: 0;
  position: relative;

  &__helper {
    content: ' ';
    display: block;
    position: absolute;
    // background-color: rgba(255,0,0,1);

    &--vertical {
      width: 1px;
      left: -1px;
      top: -100vh;
      height: 200vh;
    }

    &--horizontal {
      height: 1px;
      left: -100vh;
      width: 200vh;
    }
  }


  &--rotated {
    transform: rotate(-90deg);
  }

  &__handle {
    position: absolute;
    display: flex;
    justify-content: center;
    align-items: center;
    border: 1px solid black;
    box-sizing: border-box;
    top: -8px;
    left: -8px;
    width: 16px;
    height: 16px;
    background-color: red;
    border-radius: 50%;
    cursor: grab;
    transition: all 0.25s;
    z-index: 1001;

    &:active {
      cursor: grabbing;
    }

    &--active {
      width: 24px;
      height: 24px;
      top: -12px;
      left: -12px;
    }
  }
}
</style>
