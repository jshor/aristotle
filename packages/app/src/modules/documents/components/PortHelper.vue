<template>
  <div
    class="port-helper snappable"
    :class="`port-helper--${orientation}`"
  />
</template>

<script lang="ts">
import Vue from 'vue'
import { Component, Prop } from 'vue-property-decorator'

@Component({})
export default class PortHelper extends Vue {
  @Prop({ default: 'horizontal' })
  /**
   * Sets the visual orientation for the port snap helper.
   *
   * @type {string}
   */
  public orientation: string
}
</script>

<style lang="scss">
.port-helper {
  content: ' ';
  display: block;
  position: absolute;
  background-color: red;

  &--vertical {
    width: 1px;
    left: -1px;
    top: -200vh;
    height: 400vh;
  }

  &--horizontal {
    height: 1px;
    left: -200vh;
    width: 400vh;
  }
}
</style>
