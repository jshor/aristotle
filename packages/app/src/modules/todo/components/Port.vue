<template>
  <div class="port" :title="rotation" :style="{
    transform: `rotate(-${rotation * 90}deg)`
  }">
    <div class="port__handle" :class="{
    'port__handle--active': active
  }">
      <draggable class="port snappable" @drag="onDrag" @dragStart="dragStart" @dragEnd="dragEnd" :position="draggablePosition" :data-id="id" snap="port--input">
        <div class="port__handle dr" />
      </draggable>
      <slot />
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { Component, Prop } from 'vue-property-decorator'
import Draggable from './Draggable.vue'

@Component({
  components: {
    Draggable
  }
})
export default class Port extends Vue {

  @Prop()
  public id!: string

  @Prop({ default: 0 })
  public rotation: any

  @Prop({ default: false })
  public active: boolean

  public cloneId: string = 'DRAGGED_PORT'

  public draggablePosition: any = {
    x: 0,
    y: 0
  }

  dragStart ({ position }) {
    const { x, y } = this.$el.getBoundingClientRect()

    this.draggablePosition = {}
    this.onDrag(position)

    this.$emit('dragStart', {
      source: this.id,
      target: this.cloneId
    })
  }

  onDrag ({ position }) {
    this.$emit('drag', {
      [this.cloneId]: {
        position,
        id: this.cloneId
      }
    })
  }

  dragEnd ({ position, snappedId }) {
    this.draggablePosition = {
      x: 0,
      y: 0
    }

    this.$emit('dragEnd', {
      source: this.id,
      oldTarget: this.cloneId,
      newTarget: snappedId
    })
  }
}
</script>

<style lang="scss">
.port {
  width: 0;
  height: 0;
  position: relative;

  &--rotated {
    transform: rotate(-90deg);
  }

  &__handle {
    position: absolute;
    display: flex;
    justify-content: center;
    align-items: center;
    top: -8px;
    left: -8px;
    width: 16px;
    height: 16px;
    background-color: red;
    border-radius: 50%;
    cursor: pointer;
    transition: all 0.25s;
  z-index: 1001;

    &--active {
      width: 32px;
      height: 32px;
      top: -16px;
      left: -16px;
    }
  }
}
</style>
