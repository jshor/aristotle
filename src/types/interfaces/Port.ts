import BinaryWavePulse from '@/store/document/oscillator/BinaryWavePulse'
import PortType from '../enums/PortType'
import Direction from '../enums/Direction'
import Point from './Point'

interface Port {
  id: string
  name: string
  elementId: string
  virtualElementId?: string
  position: Point
  type: PortType
  rotation: number
  orientation: Direction
  value: number
  hue: number
  isFreeport: boolean
  isMonitored: boolean
  connectedPortIds: string[]
  wave?: BinaryWavePulse
}

export default Port
