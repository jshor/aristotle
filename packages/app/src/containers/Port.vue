<template>
  <port-pivot :rotation="rotation">
    <div v-if="showHelper">
      <port-helper orientation="horizontal" />
      <port-helper orientation="vertical" />
    </div>

    <draggable
      v-if="draggable"
      :position="draggablePosition"
      :data-id="id"
      :class="{ snappable }"
      snap=".port"
      @drag="onDrag"
      @dragStart="dragStart"
      @dragEnd="dragEnd"
      class="port">
      <port-handle
        :active="snappable"
        @mousedown="mousedown"
        @mouseup="mouseup"
      />
    </draggable>
    <port-handle
      v-else
      :active="snappable"
      @mousedown="mousedown"
      @mouseup="mouseup"
    />
  </port-pivot>
</template>

<script lang="ts">
import { defineComponent, PropType } from 'vue'
import { mapActions, mapGetters, mapState } from 'vuex'
import IPoint from '../interfaces/IPoint'
import Draggable from './Draggable.vue'
import PortHandle from '../components/PortHandle.vue'
import PortHelper from '../components/PortHelper.vue'
import PortPivot from '../components/PortPivot.vue'
import getAncestor from '../utils/getAncestor'

export default defineComponent({
  components: {
    Draggable,
    PortHandle,
    PortHelper,
    PortPivot
  },
  props: {
    /**
     * Port ID.
     */
    id: String,

    /**
     * List of sibling ports (all ports on the parent element).
     */
    siblings: {
      type: Object,
      default: () => ({})
    },

    /**
     * Unit circle rotation value (0, 1, 2, 3).
     */
    rotation: {
      type: Number,
      default: 0
    },

    /**
     * Position of the port on the canvas.
     */
    position: {
      type: Object as PropType<IPoint>,
      default: () => ({
        x: 0,
        y: 0
      })
    },

    /**
     * Whether or not the port is draggable.
     */
    draggable: {
      type: Boolean,
      default: false
    },

    /**
     * Whether or not the snap helpers should be enabled.
     * Should be set to true when the port on the opposite end of a connection is being dragged.
     */
    showHelper: {
      type: Boolean,
      default: false
    },

    /**
     * The port type.
     * - 0 = output
     * - 1 = input
     * - 2 = freeport
     */
    type: {
      type: Number,
      default: 0
    },

    /**
     * Port directional orientation.
     */
    orientation: {
      type: Number,
      default: 0
    }
  },
  data () {
    return {
      /**
       * The temporary ID of the draggable port.
       */
      cloneId: 'DRAGGED_PORT',

      /**
       * The relative dragged position of the dragged port.
       */
      draggablePosition: {
        x: 0,
        y: 0
      } as IPoint,

      /**
       * The absolute position of the dragged port.
       */
      absolutePosition: {
        x: 0,
        y: 0
      } as IPoint
    }
  },
  computed: {
    ...mapGetters([
      'zoom'
    ]),
    ...mapState([
      'activePort'
    ]),

    /**
     * Whether or not this port can accept an incoming dragged port.
     * This will update the UI to make the port bigger when true.
     */
    snappable () {
      return this.activePort !== null // && activePort.type !== this.type
    }
  },
  mounted () {
    window.addEventListener('mouseup', this.mouseup)
  },
  destroy () {
    window.removeEventListener('mouseup', this.mouseup)
  },
  methods: {
    ...mapActions([
      'updatePortPositions',
      'setActivePort',
      'connect',
      'disconnect',
      'showPortSnapHelpers',
      'hidePortSnapHelpers'
    ]),

    /**
     * Sets the absolute position of the static (i.e., not dragged) port.
     */
    setAbsolutePosition () {
      const { x, y } = this.$el.getBoundingClientRect()
      const editor = getAncestor(this, 'canvas')

      this.absolutePosition = editor.fromDocumentToEditorCoordinates({ x, y })
    },

    /**
     * Tells the store to show the snap helpers of ports adjacent to this one.
     */
    mousedown () {
      if (this.type === 2) {
        const siblingIds = Object
          .values(this.siblings)
          .reduce((ids: string[], s: any) => [
            ...ids,
            ...s.map(({ id }) => id)
          ], [])

        this.showPortSnapHelpers(siblingIds)
      }
    },

    /**
     * Tells the store that the port's no longer active, and to hide the helpers.
     */
    mouseup () {
      this.setActivePort(null)
      this.hidePortSnapHelpers()
    },

    /**
     * Initializes the draggable connection.
     */
    dragStart ({ position }) {
      this.setAbsolutePosition()
      this.setActivePort({
        id: this.cloneId,
        position: {
          x: 0,
          y: 0
        },
        type: this.type,
        orientation: this.orientation
      })

      this.draggablePosition = {
        x: NaN,
        y: NaN
      }
      this.onDrag(position)

      this.connect({
        source: this.id,
        target: this.cloneId
      })
    },

    /**
     * Updates the store position with the absolute position of the port.
     */
    onDrag ({ position }) {
      if (!position) return
      this.updatePortPositions({
        [this.cloneId]: {
          position: {
            // port coordinates are stored absolutely (relative to its existing position)
            x: position.x + this.absolutePosition.x,
            y: position.y + this.absolutePosition.y
          },
          id: this.cloneId,
          type: this.type,
          orientation: this.orientation,
          rotation: 0
        }
      })
    },

    /**
     * Handles terminating the temporary active port, and connects the port to the one at its dragged location.
     * This method handles all results of a user-driven port-dragging interaction.
     */
    dragEnd ({ position, snappedId }) {
      this.draggablePosition = {
        x: 0,
        y: 0
      }
      this.disconnect({
        source: this.id,
        target: this.cloneId
      })
      this.connect({
        source: this.id,
        target: snappedId
      })
      this.setActivePort(null)
    }
  }
})
</script>
