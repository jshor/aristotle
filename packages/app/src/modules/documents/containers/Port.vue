<template>
  <port-pivot :rotation="rotation">
    <div v-if="showHelper">
      <port-helper orientation="horizontal" />
      <port-helper orientation="vertical" />
    </div>

    <draggable
      v-if="draggable"
      :position="draggablePosition"
      :data-id="id"
      :class="{ snappable }"
      snap=".port"
      @drag="onDrag"
      @dragStart="dragStart"
      @dragEnd="dragEnd"
      class="port">
      <port-handle
        :active="snappable"
        @mousedown="mousedown"
        @mouseup="mouseup"
      />
    </draggable>
    <port-handle
      v-else
      :active="snappable"
      @mousedown="mousedown"
      @mouseup="mouseup"
    />
  </port-pivot>
</template>

<script lang="ts">
import Vue from 'vue'
import { Component, Prop } from 'vue-property-decorator'
import Draggable from './Draggable.vue'
import PortHandle from '../components/PortHandle.vue'
import PortHelper from '../components/PortHelper.vue'
import PortPivot from '../components/PortPivot.vue'
import { Getter } from '../store/decorators'
import IPoint from '../../../interfaces/IPoint'
import getAncestor from '../../../utils/getAncestor'
import Canvas from './Documents.vue'

@Component({
  components: {
    Draggable,
    PortHandle,
    PortHelper,
    PortPivot
  }
})
export default class Port extends Vue {
  /**
   * Current zoom value of the editor.
   *
   * @type {number}
   */
  @Getter('documents', 'zoom')
  public zoom: number

  /**
   * Port ID.
   *
   * @type {string}
   */
  @Prop()
  public id: string

  /**
   * List of sibling ports (all ports on the parent element).
   *
   * @type {IPort[]}
   */
  @Prop({ default: () => ({}) })
  public siblings: any[]

  /**
   * Unit circle rotation value (0, 1, 2, 3).
   *
   * @type {number}
   */
  @Prop({ default: 0 })
  public rotation: any

  /**
   * Position of the port on the canvas.
   *
   * @type {IPoint}
   */
  @Prop({
    default: () => ({
      x: 0,
      y: 0
    })
  })
  public position: IPoint

  /**
   * Whether or not the port is draggable.
   *
   * @type {boolean}
   */
  @Prop({ default: true })
  public draggable: boolean

  /**
   * Whether or not the snap helpers should be enabled.
   * Should be set to true when the port on the opposite end of a connection is being dragged.
   *
   * @type {boolean}
   */
  @Prop({ default: false })
  public showHelper: boolean

  /**
   * The port type.
   * - 0 = output
   * - 1 = input
   * - 2 = freeport
   *
   * @type {number}
   */
  @Prop({ default: 0 })
  public type: number

  /**
   * Port directional orientation.
   *
   * @type {number}
   */
  @Prop({ default: 0 })
  public orientation: number

  /**
   * The temporary ID of the draggable port.
   *
   * @type {string}
   */
  public cloneId: string = 'DRAGGED_PORT'

  /**
   * The relative dragged position of the dragged port.
   *
   * @type {IPoint}
   */
  public draggablePosition: IPoint = {
    x: 0,
    y: 0
  }

  /**
   * The absolute position of the dragged port.
   *
   * @type {IPoint}
   */
  public absolutePosition: IPoint = {
    x: 0,
    y: 0
  }

  /**
   * Whether or not this port can accept an incoming dragged port.
   * This will update the UI to make the port bigger when true.
   *
   * @type {boolean}
   */
  get snappable () {
    const { activePort } = this.$store.state.documents

    return activePort !== null // && activePort.type !== this.type
  }

  mounted () {
    window.addEventListener('mouseup', this.mouseup)
  }

  destroy () {
    window.removeEventListener('mouseup', this.mouseup)
  }

  /**
   * Sets the absolute position of the static (i.e., not dragged) port.
   */
  setAbsolutePosition () {
    const { x, y } = this.$el.getBoundingClientRect()
    const editor = getAncestor(this, Canvas) as Canvas

    this.absolutePosition = editor.fromDocumentToEditorCoordinates({ x, y })
  }

  /**
   * Tells the store to show the snap helpers of ports adjacent to this one.
   */
  mousedown () {
    if (this.type === 2) {
      const siblingIds = Object
        .values(this.siblings)
        .reduce((ids, s) => [
          ...ids,
          ...s.map(({ id }) => id)
        ], [])

      this.$store.dispatch('showPortSnapHelpers', siblingIds)
    }
  }

  /**
   * Tells the store that the port's no longer active, and to hide the helpers.
   */
  mouseup () {
    this.$store.dispatch('setActivePort', null)
    this.$store.dispatch('hidePortSnapHelpers')
  }

  /**
   * Initializes the draggable connection.
   *
   * @param {}
   */
  dragStart ({ position }) {
    this.setAbsolutePosition()
    this.setActivePort({
      id: this.cloneId,
      position: {
        x: 0,
        y: 0
      },
      type: this.type,
      orientation: this.orientation
    })

    this.draggablePosition = {
      x: NaN,
      y: NaN
    }
    this.onDrag(position)

    this.$store.dispatch('connect', {
      source: this.id,
      target: this.cloneId
    })
  }

  /**
   * Updates the store position with the absolute position of the port.
   *
   * @param {*}
   */
  onDrag ({ position }) {
    if (!position) return
    this.$store.dispatch('updatePortPositions', {
      [this.cloneId]: {
        position: {
          // port coordinates are stored absolutely (relative to its existing position)
          x: position.x + this.absolutePosition.x,
          y: position.y + this.absolutePosition.y
        },
        id: this.cloneId,
        type: this.type,
        orientation: this.orientation,
        rotation: 0
      }
    })
  }

  /**
   * Handles terminating the temporary active port, and connects the port to the one at its dragged location.
   * This method handles all results of a user-driven port-dragging interaction.
   *
   * @param {*}
   */
  dragEnd ({ position, snappedId }) {
    this.draggablePosition = {
      x: 0,
      y: 0
    }
    this.$store.dispatch('disconnect', {
      source: this.id,
      target: this.cloneId
    })
    this.$store.dispatch('connect', {
      source: this.id,
      target: snappedId
    })
    this.$store.dispatch('setActivePort', null)
  }

  /**
   * Tells the store which port is being actively-dragged.
   *
   * @param {Port} port
   */
  setActivePort (port) {
    this.$store.dispatch('setActivePort', port)
  }
}
</script>
