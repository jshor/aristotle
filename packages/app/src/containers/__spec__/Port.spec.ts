import { mount } from '@vue/test-utils'
import { createStore } from 'vuex'
import Port from '../Port.vue'
import PortHelper from '../../components/PortHelper.vue'

describe('Port Container', () => {
  let wrapper, state

  const createWrapper = (props = {}) => {
    state = {
      documents: {
        activePort: null
      }
    }

    return mount(Port, {
      props: {
        siblings: {
          left: [],
          top: [],
          bottom: [],
          right: []
        },
        ...props
      },
      global: {
        plugins: [
          createStore({
            state,
            getters: {
              zoom: () => 1
            },
            actions: {
              connect: jest.fn(),
              disconnect: jest.fn(),
              setActivePort: jest.fn(),
              hidePortSnapHelpers: jest.fn(),
              showPortSnapHelpers: jest.fn(),
              updatePortPositions: jest.fn()
            }
          })
        ]
      }
    })
  }

  beforeEach(() => {
    wrapper = createWrapper()
  })

  afterEach(() => jest.resetAllMocks())

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should show the snap helpers when the `showHelper` is true', () => {
    wrapper = createWrapper({ showHelper: true })

    const helpers = wrapper.findAllComponents(PortHelper)

    expect(helpers).toHaveLength(2)
    expect(helpers.at(0).props().orientation).toEqual('horizontal')
    expect(helpers.at(1).props().orientation).toEqual('vertical')
  })

  it('should hide the snap helpers when the `showHelper` is false', () => {
    wrapper = createWrapper({ showHelper: false })

    expect(wrapper.findAllComponents(PortHelper)).toHaveLength(0)
  })

  describe('setAbsolutePosition()', () => {
    it('should define `absolutePosition` to be the editor coordinates of the element\'s location', () => {
      expect(wrapper.vm.absolutePosition).toEqual({ x: 0, y: 0 })
      wrapper.vm.setAbsolutePosition()
      expect(wrapper.vm.absolutePosition).toEqual({ x: 10, y: 20 })
    })
  })

  describe('mousedown()', () => {
    describe('when the port is a Freeport', () => {
      it('should dispatch showPortSnapHelpers with the ids of its sibling ports', () => {
        const id = 'test-1234'

        wrapper = createWrapper({
          id,
          siblings: {
            left: [],
            top: [{ id }],
            bottom: [],
            right: []
          },
          type: 2
        })

        const spy = jest.spyOn(wrapper.vm.$store, 'dispatch')

        wrapper.vm.mousedown()

        expect(spy).toHaveBeenCalledTimes(1)
        expect(spy).toHaveBeenCalledWith('showPortSnapHelpers', [id])
      })
    })

    describe('when the port is not a Freeport', () => {
      it('should not dispatch showPortSnapHelpers', () => {
        wrapper = createWrapper({ type: 1 })

        const spy = jest.spyOn(wrapper.vm.$store, 'dispatch')

        wrapper.vm.mousedown()

        expect(spy).not.toHaveBeenCalled()
      })
    })
  })


  describe('mousedown()', () => {
    let spy

    beforeEach(() => {
      spy = jest.spyOn(wrapper.vm.$store, 'dispatch')
      wrapper.vm.mouseup()
    })

    it('should tell the store to set the active port to null', () => {
      expect(spy).toHaveBeenCalledTimes(2)
      expect(spy).toHaveBeenCalledWith('setActivePort', null)
    })

    it('should tell the store to disable the port helpers', () => {
      expect(spy).toHaveBeenCalledWith('hidePortSnapHelpers')
    })
  })

  describe('dragStart()', () => {
    const id = 'test-1234'
    const type = 1
    const orientation = 3
    const position = {
      x: 34,
      y: 56
    }

    beforeEach(() => {
      wrapper = createWrapper({
        id,
        type,
        orientation
      })
    })

    it('should set the absolute position of the element', () => {
      const spy = jest.spyOn(wrapper.vm, 'setAbsolutePosition')

      wrapper.vm.dragStart({ position })

      expect(spy).toHaveBeenCalledTimes(1)
    })

    it('should create a new active port with the data of the actively-dragged one', () => {
      const spy = jest.spyOn(wrapper.vm, 'setActivePort')

      wrapper.vm.dragStart({ position })

      expect(spy).toHaveBeenCalledTimes(1)
      expect(spy).toHaveBeenCalledWith({
        id: wrapper.vm.cloneId,
        position: {
          x: 0,
          y: 0
        },
        type,
        orientation
      })
    })

    it('should set the coordinates of the dragged port to `NaN`', () => {
      wrapper.vm.dragStart({ position })

      expect(wrapper.vm.draggablePosition).toEqual({ x: NaN, y: NaN })
    })

    it('should invoke `onDrag()` with the given position', () => {
      const spy = jest.spyOn(wrapper.vm, 'onDrag').mockImplementation(jest.fn())

      wrapper.vm.dragStart({ position })

      expect(spy).toHaveBeenCalled()
      expect(spy).toHaveBeenCalledWith(position)
    })

    it('should establish a connection between the static port and the dragged (i.e., cloned) one', () => {
      const spy = jest.spyOn(wrapper.vm.$store, 'dispatch')

      wrapper.vm.dragStart({ position })

      expect(spy).toHaveBeenCalled()
      expect(spy).toHaveBeenCalledWith('connect', {
        source: id,
        target: wrapper.vm.cloneId
      })
    })
  })

  describe('onDrag()', () => {
    const id = 'test-1234'
    const type = 1
    const orientation = 3
    const position = {
      x: 34,
      y: 56
    }

    beforeEach(() => {
      wrapper = createWrapper({
        id,
        type,
        orientation
      })
    })

    describe('when the position is not defined in the payload', () => {
      it('should not make any calls to the store', () => {
        const spy = jest.spyOn(wrapper.vm.$store, 'dispatch')

        wrapper.vm.onDrag({})

        expect(spy).not.toHaveBeenCalled()
      })
    })

    describe('when the position is defined in the payload', () => {
      it('should update the store record of the port\'s position', () => {
        const spy = jest.spyOn(wrapper.vm.$store, 'dispatch')
        const { cloneId, absolutePosition } = wrapper.vm

        wrapper.vm.onDrag({ position })

        expect(spy).toHaveBeenCalledWith('updatePortPositions', {
          [cloneId]: {
            position: {
              x: position.x + absolutePosition.x,
              y: position.y + absolutePosition.y
            },
            id: cloneId,
            type,
            orientation,
            rotation: 0
          }
        })
      })
    })
  })

  describe('dragEnd()', () => {
    const snappedId = 'snapped-el-id'
    const id = 'test-port-id-1234'
    const position = {
      x: 34,
      y: 56
    }
    let spy

    beforeEach(() => {
      wrapper = createWrapper({ id })
      spy = jest.spyOn(wrapper.vm.$store, 'dispatch')

      wrapper.vm.dragEnd({ position, snappedId })
    })

    it('should reset the `draggablePosition`', () => {
      expect(wrapper.vm.draggablePosition).toEqual({ x: 0, y: 0 })
    })

    it('should disconnect the temporary-dragged port from the current one', () => {
      expect(spy).toHaveBeenCalledWith('disconnect', {
        source: id,
        target: wrapper.vm.cloneId
      })
    })

    it('should connect the port to the one that it was snapped to', () => {
      expect(spy).toHaveBeenCalledWith('connect', {
        source: id,
        target: snappedId
      })
    })

    it('should reset the active port to null', () => {
      expect(spy).toHaveBeenCalledWith('setActivePort', null)
    })
  })
})
