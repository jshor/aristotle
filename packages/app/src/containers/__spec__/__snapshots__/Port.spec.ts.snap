// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Port Container should match the snapshot 1`] = `
<div class="port-pivot" style="transform: rotate(-0deg);">
  <!--v-if-->
  <div class="port-handle port-handle--active"></div>
</div>
`;
